system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1289,215 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+  return TRUE;
+}
+
+
+static void
+revokefs_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = getpwnam("flatpak");
+  setgid (passwd->pw_gid);
+  setuid (passwd->pw_uid);
+
+  close (GPOINTER_TO_INT(user_data));
+}
+
+static void
+name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  OngoingPull *pull = (OngoingPull *) user_data;
+
+  ongoing_pull_free (pull);
+}
+
+static OngoingPull *
+ongoing_pull_new (FlatpakSystemHelper   *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *ref,
+                  const gchar           *commit,
+                  uid_t                  uid,
+                  gchar                 *revokefs_fuse_path,
+                  gchar                 *src,
+                  gchar                 *mnt,
+                  GError               **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  OngoingPull *pull;
+  g_autoptr (GSubprocessLauncher) launcher = NULL;
+  g_autoptr (GSubprocess) revokefs_backend = NULL;
+  g_autofree gchar *target_socket = NULL;
+  int sockets[2];
+
+  pull = g_slice_new0 (OngoingPull);
+  pull->object = object;
+  pull->invocation = invocation;
+  pull->ref = g_strdup (ref);
+  pull->commit = g_strdup (commit);
+  pull->src_dir = g_strdup (src);
+  pull->mountpoint = g_strdup (mnt);
+  pull->cancellable = g_cancellable_new ();
+  pull->uid = uid;
+
+  pull->watch_id = g_bus_watch_name_on_connection (connection,
+                                                   g_dbus_connection_get_unique_name (connection),
+                                                   G_BUS_NAME_WATCHER_FLAGS_NONE, NULL,
+                                                   name_vanished_cb,
+                                                   pull, NULL); /* TODO: Take a ref here */
+
+  if (socketpair (AF_UNIX, SOCK_SEQPACKET, 0, sockets) == -1)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "Syscall socketpair failed");
+      return NULL;
+    }
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_set_child_setup (launcher,
+                                         revokefs_child_setup, GINT_TO_POINTER (sockets[1]),
+                                         NULL);
+  g_subprocess_launcher_take_fd (launcher, sockets[0], 3);
+  target_socket = g_strdup ("--socket=3");
+  pull->revokefs_backend = g_subprocess_launcher_spawn (launcher,
+                                                        error,
+                                                        revokefs_fuse_path,
+                                                        "--backend",
+                                                        target_socket, src, NULL);
+  if (pull->revokefs_backend == NULL)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_SPAWN_FAILED,
+                   "Could not spawn revokefs backend");
```

Comment:
would be nicer to pass error to the spawn call

Reply:
`error` actually is passed to the spawn call. Instead, this should translate it into something that can be sent over D-Bus.

Reply:
hmm, not sure what are trying to convey here. The error is set with `_SPAWN_FAILED` code and then sent to the client over D-Bus. See error handling in case of `ongoing_pull_new` returns `NULL` with an error.

Reply:
You have passed `error` to the call to `spawn`. If spawn fails, it will have been set; it's a bug to call `g_set_error (error, ...)` if `*error != NULL`.